addons/pos_sale/static/src/app/services/pos_store.js

```diff
@@ -170,134 +186,95 @@ patch(PosStore.prototype, {
+                accountTaxHelpers.prepare_base_line_for_taxes_computation(
+                    saleOrderLine,
+                    this.prepareSoBaseLineForTaxesComputationExtraValues(saleOrder, saleOrderLine)
+                )
             );
-            const percentageBase =
-                !down_payment_tax || down_payment_tax.price_include
-                    ? sale_order.amount_unpaid
-                    : sale_order.amount_untaxed;
-            proposed_down_payment = (percentageBase * userValue) / 100;
         }
-        if (proposed_down_payment > sale_order.amount_unpaid) {
-            this.dialog.add(AlertDialog, {
-                title: _t("Error amount too high"),
-                body: _t(
-                    "You have tried to charge a down payment of %s but only %s remains to be paid, %s will be applied to the purchase order line.",
-                    this.env.utils.formatCurrency(proposed_down_payment),
-                    this.env.utils.formatCurrency(sale_order.amount_unpaid),
-                    this.env.utils.formatCurrency(sale_order.amount_unpaid || 0)
-                ),
-            });
-            proposed_down_payment = sale_order.amount_unpaid || 0;
-        }
-        this._createDownpaymentLines(sale_order, proposed_down_payment);
-    },
-    async _createDownpaymentLines(sale_order, total_down_payment) {
-        //This function will create all the downpaymentlines. We will create one downpayment line per unique tax combination
-        const percentage = total_down_payment / sale_order.amount_total;
-        const grouped = Object.groupBy(
-            sale_order.order_line.filter((ol) => ol.product_id),
-            (ol) => ol.tax_ids.map((tax_id) => tax_id.id).sort((a, b) => a - b)
-        );
-
-        // We need one unique line for the fixed amount taxes
-        let fixed_taxes_downpayment = 0;
-        const fixed_taxes_tab = [];
-        const down_payment_line_to_create = [];
+        accountTaxHelpers.add_tax_details_in_base_lines(baseLines, this.company);
+        accountTaxHelpers.round_base_lines_tax_details(baseLines, this.company);
 
-        Object.keys(grouped).forEach(async (key) => {
-            const group = grouped[key];
+        const amount = parseFloat(payload);
+        const amountType = isPercentage ? "percent" : "fixed";
+        const downPaymentProduct = this.config.down_payment_product_id;
+        const groupingFunction = (base_line) => {
+            return {
+                grouping_key: { product_id: downPaymentProduct },
+                raw_grouping_key: { product_id: downPaymentProduct.id },
+            };
+        };
+        const downPaymentBaseLines = accountTaxHelpers.prepare_down_payment_lines(
+            baseLines,
+            this.company,
+            amountType,
+            amount,
+            {
+                computation_key: "down_payment", // TODO: won't work with multiple down payment on the same order... is it a problem?
```

Comment:
It won't be a problem if before calling this method `downPaymentSO`, we ensure that the current order doesn't have any downpayment line.

Reply:
@caburj How do you ensure this? Do you remove the previous down payment in that case?

Reply:
I created a task for this. TASK-ID: 4658169